forge-ai/src/main/java/forge/ai/simulation/SimulationController.java

```diff
@@ -81,9 +81,11 @@ public void evaluateTargetChoices(SpellAbility sa, MultiTargetSelector.Targets t
     }
 
     public void doneEvaluating(Score score) {
-        if (score.value > bestScore.value) {
+        // if we're here during a deeper level this hasn't been called for the level above yet
+        // in such case we need to check that this decision has really lead to the improvement in score
```

Comment:
Doesn't this break cases where we have something like:
1. Initial score = 100
2. Decision X, score = 50
3. Decision Y, score = 150

i.e. an intermediate decision that results in a worse state that would enable a later decision that leads to a better overall result? One benefit of SimAi is it could find those paths.

But maybe I'm missing something - I think your screenshot is trying to save that the AI wastes mana regenerating its creatures when it's not needed? Not sure how that fits in.

Reply:
I'm pretty sure only `shouldRecurse()` is responsible for aborting the tree search.

So 3. will still be reached.

I can try and find a test case for that though.

Reply:
Ok so AI had _Battlewisen Aven_ in play and I gave it 3 _Mind Rot_.

it was able to chain them all correctly to reach the highest score:
- 2 times self-target and once on opponent
- exactly 7 in grave to activate Threshold

Reply:
> Ok so AI had _Battlewisen Aven_ in play and I gave it 3 _Mind Rot_.
> 
> it was able to chain them all correctly to reach the highest score:
> 
> * 2 times self-target and once on opponent
> * exactly 7 in grave to activate Threshold

Sweet!

Ok, so if it still traverses the full tree, then I guess the effect of this is that in a case like:
100 -> 50 -> 90

It won't consider the leaf node with score 90 as best score when choosing the leaf (if there are no other options), instead it would consider "don't do anything" as the best with score 50, right? 

Assuming that's what this is doing, I'm still having trouble understanding how this changes something at the top level? Since it's recursive, wouldn't either returning 50 or 90 to the top level result in the same effect - namely top level rejecting this subtree since it's best possible score is still worse than 100? Or what am I missing?

Reply:
Since SimulationController constructor gets passed the origGameScore it should never consider any of these decisions as an option imo.

The problem was the tree looked like this:

- original score (100)
  - Play Land (raised to 200)
    - Activate Regeneration ability (stays at 200 because the Evaluator doesn't account for it either way)

now when it starts to backtrack it starts comparing the chains in opposite order so it would see that the end result of 200 would be an improvement and on the level above discard the simpler only "Play Land" chain

Reply:
I see. Where do we backtrack and compare the chains in the opposite order? This code seems to be doing it in the forward order, no? (It tracks `bestScore` starting from the initial state at this level of simulation and only updates it on >  - not >=). Just wondering if we can fix the ordering when comparing or choosing a shorter sequence when two are tied...

If that's not feasible and this really is the best way to fix it, then maybe we can expand the comment to explain what `getLastDecision().initialScore.value` corresponds to and how that differs to the initial value of `bestScore` set in the constructor.

(Sorry, not trying to be difficult on this, but genuinely confused by this logic change - thank you for taking the time to explain.)

Reply:
Yea no problem

Currently it happens here:
https://github.com/Card-Forge/forge/blob/6a4bd7ebe0e14be0b85c3b1694d632fcbd9b3aa2/forge-ai/src/main/java/forge/ai/simulation/SpellAbilityPicker.java#L379

which is only reached once we're at a leaf

It might be possible to move that deeper instead to reach the same behaviour but since the current fix is no big overhead I would be wary of introducing some side effects.

There is another call to this function from `SpellAbilityChoicesIterator` though and that one could be considered to be more forward working. Maybe you were thinking of that one?

Reply:
Oh, I see the confusion. When you were talking about "reverse order", I think you meant that if there's a simulated chain `A -->B -> C`, we'll call `doneEvaluating()` on C, before we do it on B, so the order is "reversed" in that sense.

I was instead talking about at each level, that it's done in order (over the list of candidates at that level). So if we did `A -> B1`, then will do `A -> B2`, `A -> B3`, etc (and their subtrees). In that sense, we are still going in "order". I think the bug comes from the fact that we never tell the controller of the score of just `A` before we start simulating all its children. (My original assumption was that it would be told in the constructor, but that was incorrect since the controller is created once for the full simulation tree, and not at each level, so the origin score is really the original score right at the beginning).

OK, given all the above, I agree that this was a problem - although I think the fix would be a bit cleaner if there was an explicit API call to update the score before we recurse down, rather relying on inspecting the plan like you've done. I can look at doing that and expanding the comments a bit as well later, as I'm planning to work on a related bug with the Sim AI where something similar can happen when it's responding to something on the stack where it could try to counter its own spells incorrectly because it simulated past that and decided it could cast some nice sorcery-speed spells after that (even though it could have done those anyway, but the simulation started at instant speed).

Reply:
Yes that was the cause

Alright, thanks :)